src/Init/Lean/PrettyPrinter/Parenthesizer.lean

```diff
@@ -0,0 +1,444 @@
+syntax tree and the parser (as a `Lean.Expr`) that produced it. The traversal is right-to-left to cover case 2. More
+specifically, for every Pratt parser call, we store as monadic state the precedence of the (currently) last visited
+token (`lbp`), if any, and the precedence of the nested trailing Pratt parser call (`trailRbp`), if any. If `stP` is the
+state resulting from the traversal of a Pratt parser call `p rbp`, and `st` is the state of the surrounding call, we
+parenthesize if `rbp > stP.lbp` (case 1) or if `stP.trailRbp < st.lbp` (case 2).
+
+The primary traversal is over the parser `Expr`. The `visit` function takes such a parser and, if it is the application
+of a constant `c`, looks for a `[parenthesizer c]` declaration. If it exists, we run it, which might again call `visit`.
+Otherwise, we do a single step of reduction at the head of the expression (with `default` reducibility) and try again.
+If the reduction is stuck, we fail. This happens mostly at the `Parser.mk` decl, which is irreducible, when some parser
+primitive has not been handled yet. You can think of this traversal as a special-case, customized `Expr` interpreter.
+
+The traversal over the `Syntax` object is complicated by the fact that a parser does not produce exactly one syntax
+node, but an arbitrary (but constant, for each parser) amount that it pushes on top of the parser stack. This amount can
+even be zero for parsers such as `checkWsBefore`. Thus we cannot simply pass and return a `Syntax` object to and from
+`visit`. Instead, we use a `Syntax.Traverser` that allows arbitrary movement and modification inside the syntax tree.
+Our traversal invariant is that a parser interpreter should stop at the node to the *left* of all nodes its parser
+produced, except when it is already at the left-most child. This special case is not an issue in practice since if there
+is another parser to the left that produced zero nodes in this case, it should always do so, so there is no danger of
+the left-most child being processed multiple times.
+
+Ultimately, most parenthesizers are implemented via three primitives that do all the actual syntax traversal:
+`visitParenthesizable mkParen rbp` recurses on the current node and afterwards transforms it with `mkParen` if the above
+condition for `p rbp` is fulfilled. `visitToken lbp` does not recurse but updates the current `lbp` and advances one
+node to the left. `visitArgs x` executes `x` on the right-most child of the current node and then advances one node
+to the left (of the original current node).
+-/
+
+prelude
+import Init.Lean.Parser
+import Init.Lean.Elab.Command
+import Init.Lean.Delaborator
+
+namespace Lean
+namespace Syntax
+
+/--
+Represents a cursor into a syntax tree that can be read, written, and advanced down/up/left/right.
+If the `Traverser` is used linearly, updates are linear in the `Syntax` object as well.
```

Comment:
Is this data structure similar to a Zipper?

Reply:
Yes, I think it's fair to call it one. The implementation feels a bit different because of the arrays and indices, but the spirit is the same.